### PR TITLE
Check if the argument is Thread::Backtrace::Location object

### PR DESCRIPTION
**Pull request by yui-knk**
[Bug #19262]

### CODE REVIEW

**Comment by nobu** [comment]
Why not check in `location_ptr()`?

```diff
diff --git a/vm_backtrace.c b/vm_backtrace.c
index 6e38224a91b..22e77e773f3 100644
--- a/vm_backtrace.c
+++ b/vm_backtrace.c
@@ -178,7 +178,7 @@ static inline rb_backtrace_location_t *
 location_ptr(VALUE locobj)
 {
     struct valued_frame_info *vloc;
-    GetCoreDataFromValue(locobj, struct valued_frame_info, vloc);
+    TypedData_Get_Struct(locobj, struct valued_frame_info, &location_data_type, vloc);
     return vloc->loc;
 }
 
```

File: ast.c [hunk 1]
```diff
@@ -202,6 +203,10 @@ static VALUE
 node_id_for_backtrace_location(rb_execution_context_t *ec, VALUE module, VALUE location)
 {
     int node_id;
+
+    if (!rb_obj_is_kind_of(location, rb_cBacktraceLocation)) {
```
Comment (nobu) [inline ast.c:207]:
```suggestion
    if (!rb_frame_info_p(location)) {
```
as same as `ast_s_of`.

**Comment by nobu** [comment]
This function and `rb_get_iseq_from_frame_info` can be called outside `vm_backtrace.c`.
It may be safe to check the arguments.

```diff
diff --git a/vm_backtrace.c b/vm_backtrace.c
index 718c217d734..57903572426 100644
--- a/vm_backtrace.c
+++ b/vm_backtrace.c
@@ -182,6 +182,15 @@ location_ptr(VALUE locobj)
     return vloc->loc;
 }
 
+/* same as location_ptr() with checking the data type */
+static rb_backtrace_location_t *
+rb_backtrace_location_ptr(VALUE locobj)
+{
+    struct valued_frame_info *vloc;
+    TypedData_Get_Struct(locobj, struct valued_frame_info, &location_data_type, vloc);
+    return vloc->loc;
+}
+
 static int
 location_lineno(rb_backtrace_location_t *loc)
 {
@@ -339,7 +348,7 @@ int
 rb_get_node_id_from_frame_info(VALUE obj)
 {
 #ifdef USE_ISEQ_NODE_ID
-    rb_backtrace_location_t *loc = location_ptr(obj);
+    rb_backtrace_location_t *loc = rb_backtrace_location_ptr(obj);
     return location_node_id(loc);
 #else
     return -1;
@@ -349,7 +358,7 @@ rb_get_node_id_from_frame_info(VALUE obj)
 const rb_iseq_t *
 rb_get_iseq_from_frame_info(VALUE obj)
 {
-    rb_backtrace_location_t *loc = location_ptr(obj);
+    rb_backtrace_location_t *loc = rb_backtrace_location_ptr(obj);
     const rb_iseq_t *iseq = location_iseq(loc);
     return iseq;
 }
```

**Comment by yui-knk** [comment]
Checking the argument type with `TypedData_Get_Struct` in "vm_backtrace.c" creates an error message like below. I think it's better to check the argument type within `node_id_for_backtrace_location` for better error message. Anyway do not need to expose `rb_cBacktraceLocation` then I will use `rb_frame_info_p`.

```
$ ./ruby -e 'RubyVM::AbstractSyntaxTree.node_id_for_backtrace_location(1)'
`RubyGems' were not loaded.
`error_highlight' was not loaded.
`did_you_mean' was not loaded.
`syntax_suggest' was not loaded.
<internal:ast>:112:in `node_id_for_backtrace_location': wrong argument type Integer (expected frame_info) (TypeError)
	from -e:1:in `<main>'
```

**Review by nobu — APPROVED** [review]

**Pull request merged**